### ISSUE TITLE
[MNT]: Consider deprecating kwargs for set_xticklabels()

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#23162]
### Summary

In #23158 and similar cases before people use

```
ticks = ax.get_xticks()
labels = ax.get_xticklabels()
ax.set_xticks(ticks)
ax.set_xticklabels(labels, rotation=90)
```
to change only the appearance of the ticks. Note that likely the get/set on ticks is inserted here because we have a warning that `set_xticklabels` chages to a fixed formatter and should only be used with after `set_xticks`.

The fundamental problem is that this changes to a fixed locator and formatter, which is an unintended side effect if I only want to rotate labels.

It's hard to distinguish between this sort of code and legitimate `set_xticklabels` calls. I don't think we can be more helpful with runtime warnings.

### Proposed fix

A radical suggestion is to take away styling functionality from `set_xticklabels` and instead direct people to `tick_params()`. The need for `set_xticklabels` (with it's major intent to set a fixed format - mainly for categorical values) is reduced since you can pass them as x values (`plt.bar(['apples', 'peaches'], [1, 3])`). I assume there are very few cases that people want a fixed formatter and adapt the style - which would be the only legitimate use for `set_xticklabels(labels, rotation=90)`. They would need to go the extra mile and style separately (`set_xticklabels(labels); tick_params('x', rotation=90)`). But that might be worth it for removing quite a big foot canon.

**Comment by Tian-Jionglu** [comment]
@timhoffm Thanks for explanation.

**Comment by mwaskom** [comment]
`tick_params` is not currently a replacement for `set_{axis}ticklabels`, right? It doesn't accept the arbitrary text kwargs. It's a little hard to follow all of the redirection, but one common task that I can't solve through `tick_params` is changing the label alignment after rotating.

**Comment by timhoffm** [comment]
Thanks for the hint. It's indeed true that `tick_params` is not a full replacement. So the first step would be to add API (possibly in `tick_params`) to allow full customization of the (default, i.e. applying to all tick instances of an axis) label style parameters.

**Comment by mwaskom** [comment]
IMO that would be useful whether or not this footgun gets deprecated!

**Comment by story645** [comment]
> So the first step would be to add API (possibly in `tick_params`) to allow full customization of the (default, i.e. applying to all tick instances of an axis) label style parameters.

xref #20644 since I opened that for label alignment & text properties in tick_params

**Comment by timhoffm** [comment]
Note: `set_xticklabels()` is now discouraged because setting labels without positions is error-prone. https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_xticklabels.html. It is recommended to use `set_xticks(positions, labels) instead, but that internally propagates the kwargs to `set_xticklabels()` - so the issue persists.

Overall, we should move away from setting the values for individual ticks as ticks are ephemal. The question is on the right migration:

- As discussed above, `tick_params` should grow all functionality so that it can be used in all cases instead of the kwargs.
- As for kwargs, there are 3 possibilities
  - Discourage but keep (with documentation of the difference to `tick_params`)
  - Deprecate - would need a very long deprecation period as this is a commonly used API
  - Repurpose the kwargs to feed into `tick_params` instead of hard-setting individual tick properties. This would mean the following becomes equivalent:
    ```python
    ax.set_xticks(positions, labels, rotation=90)
    # same as
    ax.set_xticks(positions, labels)
    ax.tick_params('x', roation=90)
    ```
    This is a bit dangerous because it's a behavioral change. OTOH it may be more what the users actually wanted to achieve. It also would have the advantage that people would be migrated away from the individual tick properties without having to adapt their code.
